lib/integrations/awsoidc/eks_enroll_clusters.go

```diff
@@ -0,0 +1,514 @@
+	tokenName, err := utils.CryptoRandomHex(auth.TokenLenBytes)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+	expires := clock.Now().UTC().Add(defaults.NodeJoinTokenTTL)
+
+	resourceId := uuid.NewString()
+	req := types.ProvisionTokenSpecV2{
+		SuggestedLabels: types.Labels{
+			types.InternalResourceIDLabel: apiutils.Strings{resourceId},
+		},
+		Roles: []types.SystemRole{types.RoleKube, types.RoleApp, types.RoleDiscovery},
+	}
+
+	provisionToken, err := types.NewProvisionTokenFromSpec(tokenName, expires, req)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+
+	err = tokenCreator(ctx, provisionToken)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+
+	return provisionToken.GetName(), resourceId, trace.Wrap(err)
+}
+
+func (d *defaultEnrollEKSClustersClient) InstallKubeAgent(ctx context.Context, eksCluster *eksTypes.Cluster, proxyAddr, joinToken, resourceId string, clientGetter genericclioptions.RESTClientGetter, log logrus.FieldLogger, req EnrollEKSClustersRequest) error {
+	actionConfig, err := getHelmActionConfig(clientGetter, log)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return installKubeAgent(ctx, eksCluster, proxyAddr, joinToken, resourceId, actionConfig, getHelmSettings(), req)
+}
+
+func (d *defaultEnrollEKSClustersClient) CreateToken(ctx context.Context, token types.ProvisionToken) error {
+	return nil
+}
+
+type TokenCreator func(ctx context.Context, token types.ProvisionToken) error
+
+// NewEnrollEKSClustersClient returns new client that can be used to enroll EKS clusters into Teleport.
+func NewEnrollEKSClustersClient(ctx context.Context, req *AWSClientRequest, tokenCreator TokenCreator) (EnrollEKSCLusterClient, error) {
+	eksClient, err := newEKSClient(ctx, req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	stsClient, err := newSTSClient(ctx, req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clt := defaultEnrollEKSClustersClient{
+		Client:       eksClient,
+		stsClient:    stsClient,
+		tokenCreator: tokenCreator,
+	}
+
+	return &clt, nil
+}
+
+// EnrollEKSClustersRequest contains the required fields to enroll EKS cluster to Teleport.
+type EnrollEKSClustersRequest struct {
+	// Region is the AWS Region.
+	Region string
+
+	// ClusterNames is name of the EKS cluster to enroll.
+	ClusterNames []string
+
+	// EnableAppDiscovery specifies if we should enable Kubernetes App Discovery inside the enrolled EKS cluster.
+	EnableAppDiscovery bool
+
+	// EnableAutoUpgrades specifies if we should enable agent auto upgrades.
+	EnableAutoUpgrades bool
+
+	// IsCloud specifies if enrollment is done for the Teleport Cloud client.
+	IsCloud bool
+
+	// AgentVersion specifies version of the Helm chart that will be installed during enrollment.
+	AgentVersion string
+}
+
+// EnrollEKSClusters enrolls an EKS clusters into Teleport by installing teleport-kube-agent chart on the clusters.
```

Comment:
I think we should improve on this godoc.
This is main entry for the entire feature.

For example, when there are multiple clusters and one of them fails, what happens to the others?
Maybe describe the required AWS components like AccessPolicies.

It might end up describing the implementation and not just the API itself, but I think that's fine.